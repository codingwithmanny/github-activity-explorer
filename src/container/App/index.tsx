// IMPORTS
// --------------------------------------------------------
import React, { useState, useEffect } from 'react';
import moment from 'moment';
import { getGraphColor } from '../../styles/DefaultStyles';

// Actions / Requests
import { getSearchResults, getCommitHistory } from '../../utils/requests';

// Presentation Components
import Notification from '../../components/Notification';
import Sidebar from '../../components/Sidebar';
import Graph from '../../components/Graph';
import Info from '../../components/Info';

// MAIN CONTAINER COMPONENT
// --------------------------------------------------------
/**
 * Handles most logic and passes date to presentation components
 * @returns {JSX} React compoennt
 */
const AppContainer: React.FC = () => {
  // State / Props
  const [searchResults, setSearchResults] = useState<Array<any> | null>(null);
  const [searchLoading, setSearchLoading] = useState<boolean>(false);
  const [repos, setRepos] = useState<Array<string>>([]);
  const [history, setHistory] = useState<Array<any>>([]);
  const [historyLoading, setHistoryLoading] = useState<boolean>(false);
  const [error, setError] = useState<string | null>(null);
  const [colors, setColors] = useState<Array<number>>(
    Array.apply(null, Array(52)).map((_i, k) => k + 1),
  );
  const [showDetails, setShowDetails] = useState<boolean | null>(null);

  // Functions
  /**
   * Expected graph data and formats within the week object given
   * @param {array} data Array of GitHub commits
   * @param {object} allWeeks object which contains all 52 weeks
   * @returns {object} with all data merged with allWeeks
   */
  const formatGraphData = (data: Array<any>, allWeeks: any) => {
    if (data?.length > 0) {
      return data.reduce((a: any, b: any) => {
        const date = moment(b?.commit?.author?.date);
        const key = `${
          date.month() === 11 && date.weeks() === 1
            ? date.year() + 1
            : date.year()
        }-${date.weeks()}`;

        if (a[key]) {
          a[key].commits.push({
            ...b,
            datetime: date,
          });
        }
        return a;
      }, allWeeks);
    }
    return {};
  };

  /**
   * Formats, sends request for commit history, and updates local state
   * @param {string} query repo to be queried
   * @param {object} repo GitHub repo object
   * @param {number} colorUsed index of color used to avoid duplication
   * @returns {void}
   */
  const handleCommitHistoryRequest = async (
    query: string,
    repo: any,
    colorUsed: number,
  ) => {
    setHistoryLoading(true);

    try {
      // To + From Dates
      let to = moment();

      // Adjust for monday as first day of the week
      // 0 = Sunday
      // 1 = Monday
      // ...
      const dayOfWeek = to.day();
      if (dayOfWeek !== 1) {
        if (dayOfWeek === 0) {
          to = to.clone().subtract(6, 'days');
        } else {
          to = to.clone().subtract(dayOfWeek - 1, 'days');
        }
      }
      // Get the date of 52 weeks ago
      let from = to.clone().subtract(52, 'week');

      // Http request
      const response = await getCommitHistory(
        query,
        from.format('YYYY-MM-DDTHH:MM:SSZ'),
        to.format('YYYY-MM-DDTHH:MM:SSZ'),
      );

      // Format an object with all weeks as key names
      const allWeeks = Array.apply(null, Array(53)) // 52 Weeks + 1 Current Week
        .map(() => {
          // Adjust for week number in between two years
          const key = `${
            from.month() === 11 && from.weeks() === 1
              ? from.year() + 1
              : from.year()
          }-${from.weeks()}`;

          const dateKey = [key, from.format('MMM DD, YYYY')];
          from = from.clone().add(1, 'week');
          return dateKey;
        })
        .reduce((a: any, b: any) => {
          // Create object
          a[b[0]] = {
            date: b[1],
            datetime: moment(new Date(b[1]).toISOString()),
            commits: [],
          };
          return a;
        }, {});

      // Get all logs / commit data
      const getLogs = formatGraphData(response?.data ?? [], allWeeks);

      // Format for storing data for one repo
      const data = {
        id: query,
        logs: getLogs,
        color: repo.color,
        colorUsed,
        active: true,
      };

      setColors(colors.filter((i) => i !== colorUsed));
      setHistory([...history, ...[data]]);
      setHistoryLoading(false);
      setRepos([...repos, ...[repo]]);
    } catch (error) {
      setError(
        error?.response?.data?.message ??
          error?.message ??
          'Something went wrong.',
      );
      setHistoryLoading(false);

      // Clean up remove item from selected
      const repoName = query.split('/');
      repos.find((i: any) => {
        return i?.owner?.login === repoName[0] && i?.name === repoName[1];
      });
    }
  };

  /**
   * Handles searching for repos with a query string and displays the query results
   * @param {string} query string to be searched
   * @returns {void}
   */
  const handleSearchRepo = async (query: string) => {
    setSearchLoading(true);

    if (query?.length > 0) {
      try {
        const response = await getSearchResults(query);

        if (!response?.data?.items) {
          throw new Error('Could not retrieve data.');
        }

        setSearchResults(response?.data?.items ?? []);
        setSearchLoading(false);
      } catch (error) {
        setError(
          error?.response?.data?.message ??
            error?.message ??
            'Something went wrong.',
        );
        setSearchLoading(false);
      }
    } else {
      // Reset values for blank query
      setSearchResults(null);
      setSearchLoading(false);
    }
  };

  /**
   * When a repo is selected perform the request to retrieve its commit history
   * @param {object} data Repo object data from GitHub
   * @returns {void}
   */
  const handleRepoSelected = (data: any) => {
    setSearchResults(null);
    setSearchLoading(false);

    if (!repos.find((i: any) => i?.id === data?.id) && repos.length < 50) {
      // prevent add`ing the same data / value
      const color = `${getGraphColor(colors[0])}`;

      // Get history dat
      handleCommitHistoryRequest(
        `${data?.owner?.login}/${data.name}`,
        {
          ...data,
          active: true,
          color,
        },
        colors[0],
      );
    } else {
      setError('Repo limit reached. Try removing one.');
    }
  };

  /**
   * When an repo is removed from the current selection
   * @param {object} data GitHub repo object
   * @returns {void}
   */
  const handleRemoveRepo = (data: any) => {
    // Remove from selections
    const index = repos?.findIndex((i: any) => i?.id === data?.id);

    if (index > -1) {
      setRepos([...repos.slice(0, index), ...repos.slice(index + 1)]);
    }

    // Remove from commit history
    const historyIndex = history?.findIndex(
      (i: any) => i?.id === `${data?.owner.login}/${data?.name}`,
    );
    const repo = history?.[index] ?? null;

    if (historyIndex > -1) {
      setColors([...[repo.colorUsed, ...colors]]);
      setHistory([
        ...history.slice(0, historyIndex),
        ...history.slice(historyIndex + 1),
      ]);
    }
  };

  /**
   * Toggles showing / hiding a selection
   * @param {object} data GitHub repo object
   * @returns {void}
   */
  const handleToggleShow = (data: any) => {
    const index = repos?.findIndex((i: any) => i?.id === data?.id);
    const newRepos: Array<any> = [...repos];

    if (index > -1) {
      newRepos[index].active = !newRepos[index].active;
      setRepos(newRepos);
    }

    // Update history
    const historyIndex = history?.findIndex(
      (i: any) => i?.id === `${data?.owner.login}/${data?.name}`,
    );

    if (historyIndex > -1) {
      const newHistory = [...history];
      newHistory[historyIndex].active = !newHistory[historyIndex].active;
      setHistory(newHistory);
    }
  };

  /**
   * Toggles showing / hiding app information
   * @returns {void}
   */
  const handleInfo = () => {
    setShowDetails(showDetails === null || !showDetails);
  };

  // Hooks
  /**
   * Handles when new error shows and auto hides it after 2 seconds
   */
  useEffect(() => {
    let timer: any = null;
    if (error) {
      timer = setTimeout(() => {
        setError(null);
      }, 2000);
    }
    return () => {
      clearTimeout(timer);
    };
  }, [error]);

  // Render
  return (
    <>
      <Notification message={error} />
      <Graph loading={historyLoading} history={history} />
      <Sidebar
        results={searchResults}
        loading={searchLoading}
        repos={repos}
        handleSearch={handleSearchRepo}
        handleSelected={handleRepoSelected}
        handleRemove={handleRemoveRepo}
        handleToggleSelected={handleToggleShow}
        handleInfo={handleInfo}
      />
      <Info show={showDetails} handleClose={handleInfo} />
    </>
  );
};

// EXPORTS
// --------------------------------------------------------
export default AppContainer;
