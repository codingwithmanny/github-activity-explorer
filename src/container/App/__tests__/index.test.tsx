// IMPORTS
// --------------------------------------------------------
import React from 'react';
import renderer from 'react-test-renderer';

// TESTING
// --------------------------------------------------------
import App from '../';

// TESTS
// --------------------------------------------------------
/**
 * Renders <App /> default state
 */
test('Renders App Default State', () => {
  const tree = renderer.create(<App />).toJSON();
  expect(tree).toMatchSnapshot();
});
