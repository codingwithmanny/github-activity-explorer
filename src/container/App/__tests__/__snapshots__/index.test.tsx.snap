// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders App Default State 1`] = `
Array [
  <div
    className="sc-AxjAm dXYzTG"
  >
    <div />
  </div>,
  <div
    className="sc-Axmtr chURBd"
  >
    <div
      className="sc-fzoLsD jJGyxE"
      loading="false"
    >
      <svg
        height="38"
        stroke="#fff"
        viewBox="0 0 38 38"
        width="38"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          fill="none"
          fillRule="evenodd"
        >
          <g
            strokeWidth="2"
            transform="translate(1 1)"
          >
            <circle
              cx="18"
              cy="18"
              r="18"
              strokeOpacity=".5"
            />
            <path
              d="M36 18c0-9.94-8.06-18-18-18"
            >
              <animateTransform
                attributeName="transform"
                dur="1s"
                from="0 18 18"
                repeatCount="indefinite"
                to="360 18 18"
                type="rotate"
              />
            </path>
          </g>
        </g>
      </svg>
    </div>
    <div
      className="sc-fzozJi hahaLz"
    >
      <div>
        <svg
          aria-hidden="true"
          className="StyledIconBase-ea9ulj-0 jVnGjM"
          fill="none"
          focusable="false"
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth={2}
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx={12}
            cy={12}
            r={4}
          />
          <line
            x1={1.05}
            x2={7}
            y1={12}
            y2={12}
          />
          <line
            x1={17.01}
            x2={22.96}
            y1={12}
            y2={12}
          />
        </svg>
        <p>
          Add a GitHub repo from the 
          <span>
            Search
          </span>
        </p>
      </div>
    </div>
    <div
      className="sc-AxmLO bItgTV"
      id="graph-container"
    >
      <div />
    </div>
  </div>,
  <div
    className="sc-AxhCb eLCFXh"
  >
    <button
      className="sc-AxirZ ktnnZK"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="StyledIconBase-ea9ulj-0 jVnGjM"
        fill="none"
        focusable="false"
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth={2}
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <line
          x1={5}
          x2={19}
          y1={12}
          y2={12}
        />
        <polyline
          points="12 5 19 12 12 19"
        />
      </svg>
    </button>
    <div
      className="sc-AxhUy gIZrpB"
      loading="false"
      onBlur={[Function]}
      onFocus={[Function]}
    >
      <input
        onChange={[Function]}
        placeholder="Search a GitHub Repository..."
        tabIndex={0}
        type="search"
        value=""
      />
      <svg
        aria-hidden="true"
        className="StyledIconBase-ea9ulj-0 jVnGjM"
        fill="none"
        focusable="false"
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth={2}
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx={11}
          cy={11}
          r={8}
        />
        <line
          x1={21}
          x2={16.65}
          y1={21}
          y2={16.65}
        />
      </svg>
      <ul />
    </div>
    <div
      className="sc-AxgMl hFhLOJ"
    >
      <aside>
        <svg
          aria-hidden="true"
          className="StyledIconBase-ea9ulj-0 jVnGjM"
          fill="none"
          focusable="false"
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth={2}
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx={11}
            cy={11}
            r={8}
          />
          <line
            x1={21}
            x2={16.65}
            y1={21}
            y2={16.65}
          />
        </svg>
        <p>
          <span>
            Search for a GitHub repository
          </span>
           to populate graph
        </p>
      </aside>
    </div>
    <button
      className="sc-AxiKw kYegsT"
      onClick={[Function]}
      tabIndex={0}
    >
      <svg
        aria-hidden="true"
        className="StyledIconBase-ea9ulj-0 jVnGjM"
        fill="none"
        focusable="false"
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth={2}
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx={12}
          cy={12}
          r={10}
        />
        <path
          d="M9.09 9a3 3 0 015.83 1c0 2-3 3-3 3"
        />
        <line
          x1={12}
          x2={12.01}
          y1={17}
          y2={17}
        />
      </svg>
    </button>
  </div>,
  <div
    className="sc-fzplWN knkPpE"
  >
    <div>
      <button
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="StyledIconBase-ea9ulj-0 jVnGjM"
          fill="none"
          focusable="false"
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth={2}
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            x1={18}
            x2={6}
            y1={6}
            y2={18}
          />
          <line
            x1={6}
            x2={18}
            y1={6}
            y2={18}
          />
        </svg>
      </button>
      <p>
        <small>
          <svg
            aria-hidden="true"
            className="StyledIconBase-ea9ulj-0 jVnGjM"
            fill="none"
            focusable="false"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth={2}
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9 19c-5 1.5-5-2.5-7-3m14 6v-3.87a3.37 3.37 0 00-.94-2.61c3.14-.35 6.44-1.54 6.44-7A5.44 5.44 0 0020 4.77 5.07 5.07 0 0019.91 1S18.73.65 16 2.48a13.38 13.38 0 00-7 0C6.27.65 5.09 1 5.09 1A5.07 5.07 0 005 4.77a5.44 5.44 0 00-1.5 3.78c0 5.42 3.3 6.61 6.44 7A3.37 3.37 0 009 18.13V22"
            />
          </svg>
          GitHub URL
        </small>
      </p>
      <p>
        <a
          href="undefined"
          rel="noopener noreferrer"
          target="_blank"
        >
          <code />
        </a>
      </p>
      <p>
        <small>
          <svg
            aria-hidden="true"
            className="StyledIconBase-ea9ulj-0 jVnGjM"
            fill="none"
            focusable="false"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth={2}
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx={12}
              cy={12}
              r={4}
            />
            <line
              x1={1.05}
              x2={7}
              y1={12}
              y2={12}
            />
            <line
              x1={17.01}
              x2={22.96}
              y1={12}
              y2={12}
            />
          </svg>
          GitHub COMMIT
        </small>
      </p>
      <p>
        <a
          href="undefined/commit/undefined"
          rel="noopener noreferrer"
          target="_blank"
        >
          <code />
        </a>
      </p>
      <p>
        <small>
          <svg
            aria-hidden="true"
            className="StyledIconBase-ea9ulj-0 jVnGjM"
            fill="none"
            focusable="false"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth={2}
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="14 9 9 4 4 9"
            />
            <path
              d="M20 20h-7a4 4 0 01-4-4V4"
            />
          </svg>
          Deployment
        </small>
      </p>
      <p>
        <a
          href="undefined"
          rel="noopener noreferrer"
          target="_blank"
        >
          <code />
        </a>
      </p>
      <p>
        <small>
          <svg
            aria-hidden="true"
            className="StyledIconBase-ea9ulj-0 jVnGjM"
            fill="none"
            focusable="false"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth={2}
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M20 21v-2a4 4 0 00-4-4H8a4 4 0 00-4 4v2"
            />
            <circle
              cx={12}
              cy={7}
              r={4}
            />
          </svg>
          GitHub User
        </small>
      </p>
      <p>
        <a
          href="https://github.com/undefined"
          rel="noopener noreferrer"
          target="_blank"
        >
          <code>
            @
          </code>
        </a>
      </p>
    </div>
  </div>,
]
`;
